### ISSUE TITLE
Tests fail for the latest master build on macOS

### ISSUE THREAD

**Issue by muscar** [crystal-lang/crystal#4073]
The tests for "Code gen: debug" fail on macOS Sierra with an LLVM assertion:

```
Code gen: debug
  codegens abstract struct (#3578)Assertion failed: (!Filename.empty() && "Unable to create file without name"), function createFilePathPair, file DIBuilder.cpp, line 130.
make: *** [spec] Abort trap: 6
```

Steps to reproduce:

1. `git clone https://github.com/crystal-lang/crystal.git`
2. `cd crystal; make clean crystal spec verbose=1`

Expected output: a passing test suite.

Actual output:

```
rm -rf .build
rm -rf ./doc
rm -rf src/llvm/ext/llvm_ext.o
rm -rf src/ext/sigfault.o src/ext/libcrystal.a
c++ -c  -o src/llvm/ext/llvm_ext.o src/llvm/ext/llvm_ext.cc `/usr/local/bin/llvm-config-3.6 --cxxflags`
cc -fPIC    -c -o src/ext/sigfault.o src/ext/sigfault.c
ar -rcs src/ext/libcrystal.a src/ext/sigfault.o
CRYSTAL_CONFIG_PATH=`pwd`/src ./bin/crystal build  -o .build/crystal src/compiler/crystal.cr -D without_openssl -D without_zlib
./bin/crystal build  -o .build/all_spec spec/all_spec.cr
Using compiled compiler at .build/crystal
.build/all_spec -v
Compiler
  compiles a file
  runs subcommand in preference to a filename
Code gen: alias
  invokes methods on empty array of recursive alias (1)
  invokes methods on empty array of recursive alias (2)
  invokes methods on empty array of recursive alias (3)
  casts to recursive alias
  casts to recursive alias
  doesn't break with alias for link attributes
  doesn't crash on cast to as recursive alias (#639)
  lazyly solves aliases (#1346)
  codegens cast to alias that includes bool
  overloads alias against generic (1) (#3261)
  overloads alias against generic (2) (#3261)
Code gen: and
  codegens and with bool false and false
  codegens and with bool false and true
  codegens and with bool true and true
  codegens and with bool true and false
  codegens and with bool and int 1
  codegens and with bool and int 2
  codegens and with primitive type other than bool
  codegens and with primitive type other than bool with union
  codegens and with primitive type other than bool
  codegens and with nilable as left node 1
  codegens and with nilable as left node 2
  codegens and with non-false union as left node
  codegens and with nil union as left node 1
  codegens and with nil union as left node 2
  codegens and with bool union as left node 1
  codegens and with bool union as left node 2
  codegens and with bool union as left node 3
  codegens and with bool union as left node 1
  codegens and with bool union as left node 2
  codegens and with bool union as left node 3
  codegens and with bool union as left node 4
  codegens assign in right node, after must be nilable
  codegens assign in right node, inside if must not be nil
  codegens assign in right node, after if must be nilable
Code gen: array literal spec
  creates custom non-generic array
  creates custom generic array
  creates custom generic array with type var
  creates custom generic array via alias
  creates custom generic array via alias (2)
  creates custom non-generic array in nested module
Code gen: asm
  codegens without inputs
  codegens with one input
  codegens with two inputs
Code gen: block
  generate inline
  passes yield arguments
  pass arguments to yielder function
  pass self to yielder function
  pass self and arguments to yielder function
  allows access to local variables
  can access instance vars from yielder function
  can set instance vars from yielder function
  can use instance methods from yielder function
  can call methods from block when yielder is an instance method
  nested yields
  assigns yield to argument
  can use global constant
  return from yielder function
  return from block
  return from yielder function (2)
  union value of yielder function
  allow return from function called from yielder function

  return from block that always returns from function that always yields inside if block
  return from block that always returns from function that conditionally yields
  call block from dispatch
  call block from dispatch and use local vars
  break without value returns nil
  break block with yielder inside while
  break from block returns from yielder
  break from block with value
  returns from block with value
  doesn't codegen after while that always yields and breaks
  break from block with value
  doesn't codegen call if arg yields and always breaks
  codegens nested return
  codegens nested break
  codegens call with block with call with arg that yields
  can break without value from yielder that returns nilable (1)
  can break without value from yielder that returns nilable (2)
  break with value from yielder that returns a nilable
  can use self inside a block called from dispatch
  return from block called from dispatch
  breaks from while in function called from block
  allows modifying yielded value (with literal)
  allows modifying yielded value (with variable)
  it yields nil from another call
  allows yield from dispatch call
  block with nilable type
  block with nilable type 2
  codegens block with nilable type with return (1)
  codegens block with nilable type with return (2)
  codegens block with union with return
  codegens if with call with block (ssa issue)
  codegens block with return and yield and no return
  codegens while/break inside block
  codegens block with union arg (1)
  codegens block with union arg (2)
  codegens block with virtual type arg
  codegens call with blocks of different type without args
  codegens dispatch with block and break (1)
  codegens dispatch with block and break (2)
  codegens block call when argument type changes
  returns void when called with block
  executes yield expression if no arg is given for block
  codegens bug with block and arg and var
  allows using var as block arg with outer var
  allows initialize with yield (#224)
  uses block inside array literal (bug)
  codegens method invocation on a object of a captured block with a type that was never instantiated
  codegens method invocation on a object of a captured block with a type that was never instantiated (2)
  codegens bug with yield not_nil! that is never not nil
  uses block var with same name as local var
  doesn't crash on untyped array to_s
  codegens block which always breaks but never enters (#494)
  codegens block bug with conditional next and unconditional break (1)
  codegens block bug with conditional next and unconditional break (2)
  codegens block bug with conditional next and unconditional break (3)
  codegens block bug with conditional next and unconditional break (4)
  returns from proc literal
  does next from captured block
  codegens captured block with next inside yielded block (#2097)
  codegens captured block that returns union, but proc only returns a single type
  yields inside yield (#682)
  yields splat
  yields more exps than block arg, through splat
  uses splat in block argument
  uses splat in block argument, many args
  uses block splat argument with union types
  auto-unpacks tuple
  unpacks tuple but doesn't override local variables
  codegens block with multiple underscores (#3054)
  breaks in var assignment (#3364)
  nexts in var assignment (#3364)
  dispatches with captured and non-captured block (#3969)
Code gen: c enum
  codegens enum value
  codegens enum value 2
  codegens enum value 3
  codegens enum value 4
  codegens enum with 1 + 2
  codegens enum with 3 - 2
  codegens enum with 3 * 2
  codegens enum with 10 / 2
  codegens enum with 1 << 3
  codegens enum with 100 >> 3
  codegens enum with 10 & 3
  codegens enum with 10 | 3
  codegens enum with (1 + 2) * 3
  codegens enum with 10 % 3
  codegens enum that refers to another enum constant
  codegens enum that refers to another constant
Code gen: struct
  codegens struct property default value
  codegens struct property setter
  codegens struct property setter via pointer
  codegens struct property setter via pointer
  codegens set struct value with constant
  codegens union inside struct
  codegens struct get inside struct
  codegens struct set inside struct
  codegens pointer malloc of struct
  passes struct to method (1)
  passes struct to method (2)
  codegens struct access with -> and then .
  yields struct via ->
  codegens assign struct to union
  codegens passing pointerof(struct) to fun
  builds struct setter with fun type (1)
  builds struct setter with fun type (2)
  allows using named arguments for new
  does to_s
  can access instance var from the outside (#1092)
  automatically converts numeric type in struct field assignment
  automatically converts numeric union type in struct field assignment
  automatically converts nil to pointer
  automatically converts by invoking to_unsafe
  sets instance var to proc
Code gen: c union
  codegens union property default value
  codegens union property default value 2
  codegens union property setter 1
  codegens union property setter 2
  codegens union property setter 1 via pointer
  codegens union property setter 2 via pointer
  codegens struct inside union
  codegens assign c union to union
  builds union setter with fun type
  does to_s
  automatically converts numeric type in field assignment
  automatically converts numeric union type in field assignment
  automatically converts by invoking to_unsafe
  aligns to the member with biggest align requirements
  fills union type to the max size
  reads union instance var
Code gen: case
  codegens case with one condition
  codegens case with two conditions
  codegens case with else
  codegens case that always returns
  codegens case when cond is a call
  codegens case with class
  codegens value-less case
  codegens case when constant bug (#1028)
  does case when with metaclass
Code gen: cast
  allows casting object to pointer and back
  casts from int to int
  casts from union to single type
  casts from union to single type raises TypeCastError
  casts from union to another union
  casts from union to another union raises TypeCastError
  casts from virtual to single type
  casts from virtual to single type raises TypeCastError
  casts from pointer to pointer
  casts to module
  casts from nilable to nil
  casts from nilable to nil raises TypeCastError
  casts to base class making it virtual
  casts to bigger union
  allows casting nil to Void*
  allows casting nilable type to Void* (1)
  allows casting nilable type to Void* (2)
  allows casting nilable type to Void* (3)
  casts (bug)
  can cast from Void* to virtual type (#3014)
  upcasts from non-generic to generic
  upcasts type to virtual (#3304)
  upcasts type to virtual (2) (#3304)
  casts with block var that changes type (#3341)
  casts between union types, where union has a tuple type (#3377)
  codegens class method when type id is available but not a virtual type (#3490)
  casts from nilable type to virtual type (#3512)
Code gen: class
  codegens call to same instance
  codegens instance var
  codegens recursive type
  codegens method call of instance var
  codegens new which calls initialize
  codegens method from another method without obj and accesses instance vars
  codegens virtual call that calls another method
  codgens virtual method of generic class
  changes instance variable in method (ssa bug)
  calls method on Class class
  uses number type var
  calls class method without self
  calls class method without self (2)
  assigns type to reference union type
  does to_s for class
  allows fixing an instance variable's type
  codegens initialize with instance var
  reads other instance var
  reads a virtual type instance var
  runs with nilable instance var
  runs with nil instance var when inheriting
  codegens bug #168
  allows initializing var with constant
  codegens class method
  codegens virtual class method
  allows using self in class scope
  allows using self in class scope
  makes .class always be a virtual type even if no subclasses
  does to_s for virtual metaclass type (1)
  does to_s for virtual metaclass type (2)
  does to_s for virtual metaclass type (3)
  builds generic class bug
  resolves type declaration when accessing instance var (#348)
  gets class of virtual type
  notifies superclass recursively on inheritance (#576)
  works with array in variable initializer in non-generic type (#855)
  works with array in variable initializer in generic type (#855)
  doesn't crash on instance variable assigned a proc, and never instantiated (#923)
  does to_s on class
  invokes class method inside instance method (#1119)
  codegens method of class union including Int (#1476)
  can use a Main class (#1628)
  codegens singleton (#718)
  doesn't crash if not using undefined instance variable in superclass
  codegens virtual metaclass union bug (#2597)
  doesn't crash on #1216
  doesn't crash on #1216 with pointerof
  doesn't crash on #1216 (reduced)
  doesn't crash on abstract class never instantiated (#2840)
  can assign virtual metaclass to virtual metaclass (#3007)
  transfers initializer from module to generic class
  transfers initializer from generic module to non-generic class
  transfers initializer from generic module to generic class
  doesn't skip false initializers (#3272)
  doesn't skip zero initializers (#3272)
  codegens virtual generic class instance metaclass (#3819)
Codegen: class var
  codegens class var
  codegens class var as nil
  codegens class var inside instance method
  codegens class var as nil if assigned for the first time inside method
  codegens class var inside module
  accesses class var from proc literal
  reads class var before initializing it (hoisting)
  uses var in class var initializer
  reads simple class var before another complex one
  initializes class var of union with single type
  initializes class var with array literal
  codegens second class var initializer
  initializes dependent constant before class var
  declares and initializes
  doesn't use nilable type for initializer
  codegens class var with begin and vars
  codegens class var with type declaration begin and vars
  codegens class var with nilable reference type
  initializes class var the moment it reaches it
  gets pointerof class var
  gets pointerof class var complex constant
  doesn't inherit class var value in subclass
  doesn't inherit class var value in module
  reads class var from virtual type
  reads class var from virtual type metaclass
  writes class var from virtual type
  declares var as uninitialized and initializes it unsafely
  doesn't crash with pointerof from another module
  codegens generic class class var
Code gen: closure
  codegens simple closure at global scope
  codegens simple closure in function
  codegens simple closure in function with argument
  codegens simple closure in block
  codegens closured nested in block
  codegens closured nested in block with a call with a closure with same names
  codegens closure with block that declares same var
  codegens closure with def that has an if
  codegens multiple nested blocks
  codegens closure with nested context without new closured vars
  codegens closure with nested context without new closured vars
  codegens closure with nested context without new closured vars but with block arg
  unifies types of closured var
  codegens closure with block
  codegens closure with self and var
  codegens closure with implicit self and var
  codegens closure with instance var and var
  codegens closure with instance var
  codegens closure with instance var and block
  codegen closure in instance method without self closured
  codegens closure inside initialize inside block with self
  doesn't free closure memory (bug)
  codegens nested closure
  codegens super nested closure
  codegens nested closure with block (1)
  codegens nested closure with block (2)
  codegens nested closure with nested closured variable
  codegens super nested closure with nested closured variable
  codegens proc literal with struct
  codegens closure with struct
  codegens closure with self and arguments
  codegens nested closure that mentions var in both contexts
  transforms block to proc literal
  transforms block to proc literal with free var
  allows passing block as proc literal to new and to initialize
  allows giving less block args when transforming block to proc literal
  allows passing proc literal to def that captures block with &
  allows mixing yield and block.call
  closures struct self
  doesn't form a closure if invoking class method
  doesn't form a closure if invoking class method with self
  captures block and accesses local variable (#2050)
  codegens closured self in block (#3388)
Codegen: const
  define a constant
  support nested constant
  support constant inside a def
  finds nearest constant first
  allows constants with same name
  constants with expression
  finds global constant
  define a constant in lib
  invokes block in const
  declare constants in right order
  uses correct types lookup
  codegens variable assignment in const
  declaring var
  initialize const that might raise an exception
  allows implicit self in constant, called from another class (bug)
  codegens two consts with same variable name
  works with variable declared inside if
  codegens constant that refers to another constant that is a struct
  codegens constant that is declared later because of virtual dispatch
  doesn't crash if constant is used, but class is never instantiated (#1106)
  uses const before declaring it (hoisting)
  uses const before declaring it in another module
  initializes simple const
  initializes simple const via another const
  initializes ARGC_UNSAFE
  gets pointerof constant
  gets pointerof complex constant
  gets pointerof constant inside class
Code gen: debug
  codegens abstract struct (#3578)Assertion failed: (!Filename.empty() && "Unable to create file without name"), function createFilePathPair, file DIBuilder.cpp, line 130.
make: *** [spec] Abort trap: 6
```

Additional details:

```
alex:~/A/crystal:master> uname -a
Darwin Sapiens.48hs.co 16.4.0 Darwin Kernel Version 16.4.0: Thu Dec 22 22:53:21 PST 2016; root:xnu-3789.41.3~3/RELEASE_X86_64 x86_64
alex:~/A/crystal:master> crystal -v
Crystal 0.21.0 (2017-02-21)
alex:~/A/crystal:master> /usr/local/bin/llvm-config-3.6 --version
3.6.2
```

Both the Crystal compiler and LLVM are installed using homebrew.

**Comment by muscar** [comment]
I've opened a PR wit a fix for the problem: https://github.com/crystal-lang/crystal/pull/4074.

**Comment by asterite** [comment]
Please don't use LLVM 3.6 when compiling the compiler, use LLVM 3.9

**Comment by RX14** [comment]
@asterite I would like some clarity on the llvm versions we intend to support, and the llvm versions we support at the moment.

**Comment by asterite** [comment]
We need to support LLVM 3.9. Older versions have serious bugs, like the bug we found that generates incorrect code in release mode. That's a very good reason to drop support for older LLVM versions.

**Comment by ysbaddaden** [comment]
I was about to say that the bug wasn't in 3.8 and that 3.8 is the current stable, but since 4.0 should be released in the coming weeks, then 3.9 is about to become the current stable. Yet, maybe we could still support 3.8 for some time? It's likely to stay the default version on a number of Linux distributions for some more months.

**Comment by asterite** [comment]
But we distribute LLVM with the releases, and the compiler statically link it. I think we should upgrade to 3.9 and distribute that, and stop supporting older, buggy versions.

**Comment by asterite** [comment]
I think there are some issues with LLVMContextDispose which I wasn't able to reproduce in 3.9, but happen in older versions...

**Comment by muscar** [comment]
@asterite I'll try to build it with LLVM 3.9. Can you please update https://github.com/crystal-lang/crystal/blob/master/CONTRIBUTING.md? I was under the impression that 3.6 was required for builds. As for the issues with `LLVMContextDispose`, yes, they're happening with 3.6.

**Comment by RX14** [comment]
I think we should support LLVM releases until they become hard to maintain, or we find bugs in them. For that reason increasing the minimum to LLVM 3.8 seems like a good call but what's the benefit of requiting 3.9+?

**Comment by mosop** [comment]
I've recently started to build Crystal and its applications on Alpine Linux. I believe that Crystal is very very suitable for building Docker-based microservices.

However LLVM 3.8 is buggy to compile the Crystal compiler with --release and, unfortunately, the default LLVM version is still 3.8 on Alpine. So, currently, I'm compiling the compiler with --cross-compile on Mac with LLVM 3.9 first and then link the .o file and libraries on Alpine.

Actually, I don't need you to support LLVM 3.8. I would like you to provide more official prebuilt compilers for the modern Linux distributions like Ubuntu.

In my opinion, there's no problem for most of users even if you dropped the support for LLVM 3.8. Only the contributors have to avoid the buggy LLVM versions for compiling the compiler. I would not like you to do the waste of time for fixing such a bug now.

**Comment by RX14** [comment]
LLVM 3.8 is one of the working versions actually. If you get segfaults/bugs on 3.8 please report them.

**Comment by muscar** [comment]
Compiles and works fine with llvm 3.8 on macOS for me. Feel free to close this issue. Thanks for th help.

**Comment by ysbaddaden** [comment]
@mosop I have no problem compiling Crystal in release mode with LLVM 3.8 on Alpine Linux. What error do you have?

**Comment by kirbyfan64** [comment]
@ysbaddaden Accidental close?

**Comment by ysbaddaden** [comment]
No, the initial issue is solved: use LLVM 3.8 or later.

**Comment by mosop** [comment]
@ysbaddaden Really? So, it might be my misunderstanding. I forgot how I reached that conclusion, but everything went well once I stopped to use 3.8. Anyway thank you for letting me know that.
